### ISSUE TITLE
Split Tool

### ISSUE THREAD

**Issue by PekkaSavolainen** [Spine-project/Spine-Toolbox#1356]
As discussed we decided to go with splitting the Tool into 3 (three) separate project items

1. Python and Julia Tool project item
2. Gams project item
3. Executable project item

When we have these, we can merge the functionality of the Gimlet project item into the new Executable Tool project item.

Any suggestions for the names of these new project items?

**Comment by manuelma** [comment]
Python and Julia, in my opinion, should be different project items if we go with this. I don't really see why those two in particular should be combined?

For the names, I would just go with Julia Tool, Python Tool, Gams Tool, and Executable Tool. They will all go under the Tool categories, and will have their own 'filtered' specification model.

How about the project .json file? Do we need to touch anything? Maybe not.

Also, are we sure this split is a good idea from the UI perspective? I can't say I'm 100% convinced...

**Comment by soininen** [comment]
I feel that Python and Julia Tools need to be separate, too. I'd also consider some other name than Executable Tool which sounds somewhat confusing. Maybe External Program or Shell Tool or Launcher?

**Comment by PekkaSavolainen** [comment]
Yes, let's go with separate Julia and Python project items. Makes more sense that way.

@soininen. I'm open for renaming Executable Tool to something else, but I'd like to keep `Tool` in there somehow if possible. Using `Shell` in the name though could be a bit problematic since it will also support non-shell execution when we merge Gimlet to it.

> Also, are we sure this split is a good idea from the UI perspective?

Could you (@manuelma) elaborate? What is your concern here?

**Comment by manuelma** [comment]
> Could you (@manuelma) elaborate? What is your concern here?

I don't clearly see the benefit, and it feels somewhat inconsistent. For me, this would mean that we'd also need to split Exporter and Importer across CSV, Excel, Gdx, etc, which seems awful, in order to keep it consistent. I like it the way it is, because I can create a Tool and then decide what language to use. The level of abstraction of a project item is just that, the role they have in the workflow, and Tool is a script. The language comes after. There's nothing intrinsically different in terms of the role they play in the workflow, between the different tool types. And if we split it as suggested here, it seems that Python Tools are capable of different things than Julia Tools.

Do we also want a to have a different Data Connection item depending if it holds CSVs, Excel, etc? A different Data Store item depending on whether it is SQLite, MySQL?

**Comment by manuelma** [comment]
> non-shell execution

What does this mean? What do we understand by *shell*?

**Comment by jkiviluo** [comment]
> Do we also want a to have a different Data Connection item depending if it holds CSVs, Excel, etc? A different Data Store item depending on whether it is SQLite, MySQL?

That's a valid concern. I'm trying to think it purely from the user perspective. Having those language supports visible in the tool bar is useful information. It's less important to know right away what formats you can export in. If you want to export, you'll know to drag the exporter and check what you can do. I know this is bit arbitrary and a case can be made in the other direction as well. It could even be a good thing to show the different export formats as items (but then we should think how it doesn't get too crowded - and maybe we don't want to do that effort now).

For tools, I think this boils down to the ease-of-use. When using the 'executable tool', it should be easy-to-setup and it seems like that an interface that also supports Julia/Python/GAMS would not be as clear. It may also be just about how the interface is organized. It would maybe be good to see mock-ups...

The difference between Julia/Python/GAMS and other tools is that we provide some level of direct support for them. Then there are 'other tools' that can be made work through commands or shells. I tend to agree that 'shell tool' and 'executable tool' can be bit misleading when in reality you can start a Java program with that. Maybe that 'other tool' is better?

**Comment by PekkaSavolainen** [comment]
> What do we understand by shell?

I don't know all the intricacies of shells but let me try. *cmd.exe*, *powershell.exe* and *bash* are all shells. In Gimlet properties, there's the **Shell** checkbox for convenience. If you check this, have cmd.exe selected as the shell and e.g. notepad.exe as the **command**, Toolbox executes `cmd.exe /C notepad.exe`. If you uncheck the shell box, the executed command is just `notepad.exe` (this is what I meant by non-shell execution). You can also uncheck the Shell box, and type in `cmd.exe /C notepad.exe` as **command**. When the Shell box is unchecked, toolbox executes the command given in the line edit without making any modifications to it.

But what actually happens when an .exe is executed is something I've been trying to figure out for the past couple of hours. My current understanding is that there is **GUI shell** buried in Windows called explorer.exe, so it seems that even when a user executes notepad.exe or java.exe without specifying a shell, the file is executed using a shell (explorer.exe) anyway. *But how is explorer.exe executed then?!* If this is correct then a *non-shell execution* mode does not actually exist. No idea how this works on Linux.

Anyway, when removing Gimlet and if the 'choose Shell' feature is confusing people, we should probably remove it and force users to type in the full command.

**Comment by manuelma** [comment]
Thanks for the explanation!

I never heard of non-shell execution. Programs are executed through a shell, like cmd.exe of bash, and that's what QProcess and subprocess do, isn't it?

**Comment by PekkaSavolainen** [comment]
> I never heard of non-shell execution. Programs are executed through a shell, like cmd.exe or bash, and that's what QProcess and subprocess do, isn't it?

Non-shell execution is my own terminology (sorry). On Windows, python's subprocess.Popen() prepends the command to be executed with `cmd.exe /C` **if you set `shell=True`**. If you set `shell=False` (the default) then it kind of is non-shell execution or what would you call it? QProcess does not have this `shell` argument at all, so it by default it does not use cmd.exe to run what ever program you want to run. The Shell checkbox in Gimlet properties in practice brings this shell argument to our user's fingertips. It does the very same thing than what the shell argument in subprocess.Popen() does.

**Comment by jkiviluo** [comment]
@PekkaSavolainen This is done? Close?

**Comment by PekkaSavolainen** [comment]
No, this is not done but I'm having some doubts about the whole thing. We should definitely first do #1464 and then #1349 but I don't remember the motivation on why would we split Tool into four separate project items. Sounds like refactoring for the sake of refactoring. Ok to close by my account.

**Comment by manuelma** [comment]
I agree with @PekkaSavolainen this doesn't feel like the right time to start doing this. We can write something in the road map document about it and close for now.

**Comment by jkiviluo** [comment]
Sorry, I didn't think that through. Yeah, probably not the right time at this point. Roadmap.

For me the point is to make things more clear and visible for the user. It's good to recognize from the workflow different tool categories and to see what is supported and how. I don't know if there are code-side benefits for having a clear separation (interface already changes with tool), but I guess that's also a possible benefit (or drawback).

**Comment by manuelma** [comment]
To be honest, reading through the discussion yesterday I came to think this is not a terrible idea. Might help users indeed, we just need to get it right :)
